reduce bulk size in database writer, before "Prepared statement contains too many placeholders" appears

when there are many dns requests incoming , there is no option to set the timeout for the logger ( thus flooding stdout ),  already switched from csv to mysql ..

also "Prepared statement contains too many placeholders" appears under heavy load and dumps a very long single line to stdout


log:
```
2022-11-08 20:53:35]  WARN  query_logging_resolver: query log writer is too slow, write duration: 0 ms channel_len=502
2022-11-08 20:53:35]  WARN  query_logging_resolver: query log writer is too slow, write duration: 0 ms channel_len=501
2022-11-08 20:53:45] ERROR  server: error on processing request:resolution was not successful, used resolvers: 'upstream 'https://dns.digitale-gesellschaft.ch/dns-query'' and 'upstream 'tcp-tls:fdns1.dismail.de'' errors: [can't resolve request via upstream server https://185.95.218.43:443/dns-query: http return code should be 200, but received 502 can't resolve request via upstream server [2a02:c205:3001:4558::1]:853: dial tcp [2a02:c205:3001:4558::1]:853: connect: cannot assign requested address]

2022-11-08 20:53:48]  INFO /src/querylog/database_writer.go:136 Error 1390: Prepared statement contains too many placeholders
[361.971ms] [rows:0] INSERT INTO `log_entries` (`request_ts`,`client_ip`,`client_name`,`duration_ms`,`reason`,`response_type`,`question_type`,`question_name`,`effective_tldp`,`answer`,`response_code`) VALUES ('2022-11-08 20:53:29.486','127.0.0.1','127.0.0.1',17,'RESOLVED (tcp+udp:4.2.2.4)','RESOLVED','A','facebook.com','facebook.com','A (157.240.9.35)','NOERROR'),('2022-11-08 20:53:29.502','127.0.0.1','127.0.0.1',3,'RESOLVED (tcp+udp:1.1.1.1)','RESOLVED','A','youtu.be','youtu.be','A (142.250.179.142)','NOERROR'),   ... ##  more entries like this

```

--- a/go.mod
+++ b/go.mod
@@ -36,7 +36,7 @@ require (
 )
 
 require (
-	github.com/alecthomas/template v0.0.0-20190718012654-fb15b899a751
+	github.com/DATA-DOG/go-sqlmock v1.5.0
 	github.com/docker/go-connections v0.4.0
 	github.com/dosgo/zigtool v0.0.0-20210923085854-9c6fc1d62198
 	github.com/testcontainers/testcontainers-go v0.15.0

--- a/querylog/database_writer.go
+++ b/querylog/database_writer.go
@@ -125,9 +125,7 @@ func (d *DatabaseWriter) periodicFlush() {
 
 		err := d.doDBWrite()
 
-		if err != nil {
-			log.Log().Error("can't write entries to the database: ", err)
-		}
+		util.LogOnError("can't write entries to the database: ", err)
 	}
 }
 

--- a/querylog/database_writer.go
+++ b/querylog/database_writer.go
@@ -10,6 +10,7 @@ import (
 	"gorm.io/gorm/logger"
 
 	"github.com/0xERR0R/blocky/log"
+	"github.com/hashicorp/go-multierror"
 
 	"github.com/0xERR0R/blocky/util"
 	"github.com/miekg/dns"
@@ -121,7 +122,12 @@ func (d *DatabaseWriter) periodicFlush() {
 
 	for {
 		<-ticker.C
-		d.doDBWrite()
+
+		err := d.doDBWrite()
+
+		if err != nil {
+			log.Log().Error("can't write entries to the database: ", err)
+		}
 	}
 }
 
@@ -157,16 +163,32 @@ func (d *DatabaseWriter) CleanUp() {
 	d.db.Where("request_ts < ?", deletionDate).Delete(&logEntry{})
 }
 
-func (d *DatabaseWriter) doDBWrite() {
+func (d *DatabaseWriter) doDBWrite() error {
 	d.lock.Lock()
 	defer d.lock.Unlock()
+
+	var err *multierror.Error
 
 	if len(d.pendingEntries) > 0 {
 		log.Log().Tracef("%d entries to write", len(d.pendingEntries))
 
-		// write bulk
-		d.db.Create(d.pendingEntries)
+		const bulkSize = 100
+
+		for i := 0; i < len(d.pendingEntries); i += bulkSize {
+			j := i + bulkSize
+			if j > len(d.pendingEntries) {
+				j = len(d.pendingEntries)
+			}
+			// write bulk
+			tx := d.db.Create(d.pendingEntries[i:j])
+			err = multierror.Append(err, tx.Error)
+		}
+
 		// clear the slice with pending entries
 		d.pendingEntries = nil
+
+		return err.ErrorOrNil()
 	}
+
+	return nil
 }

--- a/querylog/database_writer_test.go
+++ b/querylog/database_writer_test.go
@@ -69,7 +69,7 @@ var _ = Describe("DatabaseWriter", func() {
 				})
 
 				// force write
-				writer.doDBWrite()
+				Expect(writer.doDBWrite()).Should(Succeed())
 
 				// 2 entries in the database
 				Eventually(func() int64 {
@@ -92,6 +92,48 @@ var _ = Describe("DatabaseWriter", func() {
 
 					return res
 				}, "5s").Should(BeNumerically("==", 2))
+			})
+		})
+
+		When("> 10000 Entries were created", func() {
+			BeforeEach(func() {
+				writer, err = newDatabaseWriter(sqliteDB, 7, time.Millisecond)
+				Expect(err).Should(Succeed())
+			})
+
+			It("should be persisted in the database in bulk", func() {
+				res, err := util.NewMsgWithAnswer("example.com", 123, dns.Type(dns.TypeA), "123.124.122.122")
+				Expect(err).Should(Succeed())
+
+				response := &model.Response{
+					Res:    res,
+					Reason: "Resolved",
+					RType:  model.ResponseTypeRESOLVED,
+				}
+
+				const count = 10_123
+
+				for i := 0; i < count; i++ {
+					writer.Write(&LogEntry{
+						Request:    request,
+						Response:   response,
+						Start:      time.Now(),
+						DurationMs: 20,
+					})
+				}
+
+				// force write
+				Expect(writer.doDBWrite()).Should(Succeed())
+
+				// 2 entries in the database
+				Eventually(func() int64 {
+					var res int64
+					result := writer.db.Find(&logEntry{})
+
+					result.Count(&res)
+
+					return res
+				}, "5s").Should(BeNumerically("==", count))
 			})
 		})
 
@@ -128,7 +170,7 @@ var _ = Describe("DatabaseWriter", func() {
 				})
 
 				// force write
-				writer.doDBWrite()
+				Expect(writer.doDBWrite()).Should(Succeed())
 
 				// 2 entries in the database
 				Eventually(func() int64 {

--- a/querylog/database_writer_test.go
+++ b/querylog/database_writer_test.go
@@ -1,13 +1,18 @@
 package querylog
 
 import (
+	"database/sql"
+	"fmt"
 	"time"
 
 	"github.com/0xERR0R/blocky/log"
 	"github.com/0xERR0R/blocky/model"
 	"github.com/0xERR0R/blocky/util"
+	"github.com/DATA-DOG/go-sqlmock"
 	"github.com/miekg/dns"
 	"github.com/sirupsen/logrus"
+	"gorm.io/driver/mysql"
+	"gorm.io/driver/postgres"
 	"gorm.io/driver/sqlite"
 	"gorm.io/gorm"
 
@@ -221,6 +226,129 @@ var _ = Describe("DatabaseWriter", func() {
 				Expect(err.Error()).Should(HavePrefix("incorrect database type provided"))
 			})
 		})
+	})
+
+	Describe("Database initialization and migration", func() {
+		var (
+			db   *sql.DB
+			dlc  gorm.Dialector
+			mock sqlmock.Sqlmock
+			err  error
+		)
+
+		When("postgres database is configured", func() {
+			BeforeEach(func() {
+				db, mock, err = sqlmock.New()
+				Expect(err).Should(Succeed())
+
+				dlc = postgres.New(postgres.Config{
+					Conn: db,
+				})
+				Expect(err).Should(Succeed())
+
+				mock.MatchExpectationsInOrder(false)
+			})
+			AfterEach(func() {
+				Expect(mock.ExpectationsWereMet()).Should(Succeed())
+			})
+
+			// nolint:lll
+			It("should create the database schema automatically", func() {
+				By("create table", func() {
+					mock.ExpectExec(`CREATE TABLE "log_entries"`).WillReturnResult(sqlmock.NewResult(0, 0))
+				})
+
+				By("create indexes", func() {
+					mock.ExpectExec(`CREATE INDEX IF NOT EXISTS "idx_log_entries_response_type"`).WillReturnResult(sqlmock.NewResult(0, 0))
+					mock.ExpectExec(`CREATE INDEX IF NOT EXISTS "idx_log_entries_client_name"`).WillReturnResult(sqlmock.NewResult(0, 0))
+					mock.ExpectExec(`CREATE INDEX IF NOT EXISTS "idx_log_entries_request_ts"`).WillReturnResult(sqlmock.NewResult(0, 0))
+				})
+
+				By("create postgres specific manually defined primary key", func() {
+					mock.ExpectExec(`ALTER TABLE log_entries ADD column if not exists id serial primary key`).WillReturnResult(sqlmock.NewResult(0, 0))
+				})
+
+				_, err = newDatabaseWriter(dlc, 1, time.Millisecond)
+				Expect(err).Should(Succeed())
+			})
+		})
+
+		When("mysql database is configured", func() {
+			BeforeEach(func() {
+				db, mock, err = sqlmock.New()
+				Expect(err).Should(Succeed())
+
+				dlc = mysql.New(mysql.Config{
+					Conn:                      db,
+					SkipInitializeWithVersion: true,
+				})
+				Expect(err).Should(Succeed())
+
+				mock.MatchExpectationsInOrder(false)
+			})
+			AfterEach(func() {
+				Expect(mock.ExpectationsWereMet()).Should(Succeed())
+			})
+			// nolint:lll
+			Context("Happy path", func() {
+				It("should create the database schema automatically", func() {
+					By("create table with indexes", func() {
+						mock.ExpectExec("CREATE TABLE `log_entries`.*INDEX (`idx_log_entries_request_ts`|`idx_log_entries_client_name`|`idx_log_entries_response_type`)").WillReturnResult(sqlmock.NewResult(0, 0))
+					})
+
+					By("create mysql specific manually defined primary key", func() {
+						mock.ExpectExec("ALTER TABLE `log_entries` ADD `id` INT PRIMARY KEY AUTO_INCREMENT").WillReturnResult(sqlmock.NewResult(0, 0))
+					})
+
+					_, err = newDatabaseWriter(dlc, 1, time.Millisecond)
+					Expect(err).Should(Succeed())
+				})
+			})
+
+			// nolint:lll
+			Context("primary index creation", func() {
+				It("should create the database schema automatically without errors even if primary idex exists", func() {
+					By("create table with indexes", func() {
+						mock.ExpectExec("CREATE TABLE `log_entries`.*INDEX (`idx_log_entries_request_ts`|`idx_log_entries_client_name`|`idx_log_entries_response_type`)").WillReturnResult(sqlmock.NewResult(0, 0))
+					})
+
+					By("create mysql specific manually defined primary key should be skipped if already exists (error 1060)", func() {
+						mock.ExpectExec("ALTER TABLE `log_entries` ADD `id` INT PRIMARY KEY AUTO_INCREMENT").WillReturnError(fmt.Errorf("error 1060: duplicate column name"))
+					})
+
+					_, err = newDatabaseWriter(dlc, 1, time.Millisecond)
+					Expect(err).Should(Succeed())
+				})
+
+				It("should fail if manually defined index can't be created", func() {
+					By("create table with indexes", func() {
+						mock.ExpectExec("CREATE TABLE `log_entries`.*INDEX (`idx_log_entries_request_ts`|`idx_log_entries_client_name`|`idx_log_entries_response_type`)").WillReturnResult(sqlmock.NewResult(0, 0))
+					})
+
+					By("create mysql specific manually defined primary key should be skipped if already exists", func() {
+						mock.ExpectExec("ALTER TABLE `log_entries` ADD `id` INT PRIMARY KEY AUTO_INCREMENT").WillReturnError(fmt.Errorf("error XXX: some index error"))
+					})
+
+					_, err = newDatabaseWriter(dlc, 1, time.Millisecond)
+					Expect(err).Should(HaveOccurred())
+					Expect(err.Error()).Should(ContainSubstring("can't perform auto migration: error XXX: some index error"))
+				})
+			})
+
+			Context("table can't be created", func() {
+				It("should create the database schema automatically without errors", func() {
+					By("create table with indexes", func() {
+						mock.ExpectExec("CREATE TABLE `log_entries`").WillReturnError(fmt.Errorf("error XXX: some db error"))
+					})
+
+					_, err = newDatabaseWriter(dlc, 1, time.Millisecond)
+					Expect(err).Should(HaveOccurred())
+					Expect(err.Error()).Should(ContainSubstring("can't perform auto migration: error XXX: some db error"))
+				})
+			})
+
+		})
+
 	})
 
 })

